### DISCUSSION TITLE
A union-based version of ActorBox

### DISCUSSION THREAD

**Discussion by khoover** [uazu/stakker#39] [Show and tell]
It's been annoying me a bit that the actor state has to be held in two different places, so I [wrote up](https://rust.godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1DEArgoKkl9ZATwDKjdAGFUtEywZ7HAGTwNMAHLuAEaYxBIAnBqkAA6oCoR2DC5uHnpxCbYCvv5BLKHhXFGWmNZZDEIETMQEKe6eXCVlSZXVBDmBIWGR0eZtdWmNfTUdeQU9AJSWqCbEyOwcZpgA1OboICiltBsApADsAEJOW6TLAEqYVMe0tKcXVACyJgT7ACI7AMwHOxoAgkurAjrEBsFgbB6GExiWgAT1exFQMU%2B3z%2BP1%2BOwATB8dgBWA5YYh4ABumAgTloAkwpxcMRhp1emGCJmAp2UbTwYmwAEdTtyJrj3qisbiDsRMDFiBATAAOfk4wW/RjuZatAgrfYo37LbXLZRimKkNE686YJjoOlGnUALVQ%2BTwVLRb0dfxMDCSy1eTCqnyc6I%2B2GWGst2ol4pAywhDChNzhCKRHycADVMMgfQdUKofeaGMsAGIMADyDHmEH5/s%2B2Arhr%2BxrFZph4cj0dh8MRPr92Gr6L2CrR5lMNmWAEkGP5iO2K4HDsHlugvUxw/drrQfZ7vQmO1WZ301eHlz7VZgt6ie87fngWDEVxvJyOxxP/VPNcaqDn/AB3UvLAC0k6EpSoJ8Z2Nf9aEAoMa2NKDtR3dhlmXDYPwgQ8Nj1cUpmA6DZ3nRdLgQkAkLXJggMgrCsNDGJG0haFY0RRDME/ZNkHoz8Jgw0iyKnV52K1LCnQ4/j0Q419ljwBQAH0AC9bWCe0IExAA2KwqAmH9J2CVBXBI3joJYL1kAQTAFDASBlIAOlgszgEwAhS1OFCQBtO1MH5ATT1RYSczE8SKPkjElIA1Tf0fDStIgnSoL0ggDKMkyIHMyzrNsqYVSqXcQDQmJXIiwSZxEohxLrc0/ICsDTiJMRw1%2BVTwrI%2BgCGWFhnmwqpAw%2BV5VgAsy5yqMyNOIBF33EprkuRTDtSigzOrAiy0swKybK/WrOO1BzMra95H1dBQmCodVpw4laYK6yylFshynNklzkWWAB6W681dGwkho5ZiMxDE4jEyl0A%2BwEvSPQ6jqbGjW0okAqgAa1JRSRpapgzIo7Kjs4sy/AKwgwlLcaUeWMz%2BGIcTTQM%2BS9icKh9icZYqFLMtn1x7UeoRoqYQ2iNqJjMGWIgCraDpnGyISuaLMWhyLnrfmgegt4uxRsXTXNDbJxiQw8GQOKPt%2BGwSA2AqWa/UQbkwdBlgId81aPDEMR4uW5o2S77SVx87JxwTjVyzzTdQKSZLkxTlJqg6IuNerGuapm2fMpm%2BpIQbhuebGvhx7bdv2%2BmVsmhBlnik7hf1AwSwu32XNOWHmoAKiZwP09x5C7Yy/V6XnJ9lgo8MuRMTBO64mr2snEHObjDZ0DjCAO67kuBawuuAY2cXzSb1qNWWFnwyYbXiB7p2AwHlsh5AEfEQgdeiGIG2GezovnNOcTe82gMXalqC3b49ycrfvKc36uP9f9wK1MfAWGI5QfT3CeC8G85ZHzLV0vpLOQsAYLWStpW2s8QDz1Zp8e%2BKpbSkjAc8DYekYg5xmtHb%2BqAhojTslOJwTNKbLBTntFBF8y4NUrvOMyLMp7ajeGxWWR1xLb2WAESk/DpZv3dh/T25Chq%2BT/mBIKk4gEgITPgiBSYUxpgzFmGEOZ8xFhLGWSsUDjEBhgZFOB00qCzUQUlJaQdUHpXWlgv8uCIBqMIUwYhUcOEyPjslUuZM6FkwYQwHaTDzEo1YcsdhvUKLcJ7ufFagiXGPhEf4MRz8JE6lylI88l5ry%2BknGDamJBhyjjCA%2BMxDidQiUPsQ6JAdmHQVDrBSOudbE2X8YnGuIcbLw3aaQjhSVumS2DjqRhacEmZ2zrBU4Vdmm1zWo3D0zdl5t2WOPbuvChGRIvrvWi4N6lj07p3MZF8uIJONDPdKGDF7vUOCvBWDY3oby3qk6pvT9kcz3nRA%2Bo8T4kHOSw7JDMbnsEcsXG%2Bd9JyP3GeIhUZEX48I/nkgEawNhhAROEEA2ABokDGh5dEwo8QEmJKSBkTJgBygVDEEwwRlhKhYKlRETgmAAnMZUREMRjaZNzEwOgxsIDpkzAmbMyw8XYorEk5YABpOg9B0DCu0WK3REr8Xjn9DKsGPL0Bdg9vkq86rsWlM3lymIbKOXTjRUoQEwIqAsAILsQ4rwxJXiYHSY1ZSmAKDzI6yVpxcwkCimqccp4k5EovEa11Ch3WswJiyi17LbXmJEg62y8iqCnHTeGaJQbiAhsqQmMAHBxLSoAQGC4CgTC0HUdQ3M/qBqTj2RNSxylFlkVAlQDY5rdVCPTWZd8hI1TiX7H5DEvbjYfRlZ2gCGwdXG37Y6wdw7MCjoIJKD6C7fpWxnVhLtGx%2BWCqVS5IRQ7MZxXTaXK2R7FXLHNgQLOQZTxW1OJPJ%2BIE50gHlUbE9MLHznrVJex116MQ/rvQ%2Bp91r3ivsZXulFiKcmot7H8OlDL%2BwmEHFrU%2BIqqmLIYKgWwVAXn7gTMopIWjRW0LVfo4spJzWWqUEYisx4Ip%2BDHOGO8Rain%2Bn1XkqNhSOzLGwyQXDkDPkzjQ9TN8DEIAEaIy88jAhKM6L0YWOjddWXJuuiYxRj4u0dp1AZltUF5N4GI3uLY3MzPEfg1Bdj3RykcYIrJuzyLLlEuNFJ/K3tpLORKk04KAYlMMBU6qtTBj6NECTUsZj0CanQXMg54gZkCp%2Bauj0nG5kbMwlS0waG2M3IoYit5nMXirwwh8vqALAFs3hgE3mdThiK3LBC1U6uONzNWLRhUlL3lfIdY/TqLrochBuN5nfam9ivkrSS711L3t9YTcJfCrJiGEUzgNV5%2Bl0nlh%2BN/v5QLSjgEUdUZccBVTm0JagnNscMd8VDX1sjJDxXtsMpEn4uRh3/5Bdayd5TZ3HjPB9ExML1GIsabi6Yq7M3uvJfu3HAbm28lvd26GlgfgAY1bKrOe06xE2MdPSZkb/SxtsDk4R8zMJJvmTS8XabCScsQHwMbYFG2iuOjyVJjDg4hACcwCJze5ipPJaqhvMTTgJcyhY7xznxWGt84KQLt5wudsiSQnpgMxnrt9Iao%2BsSbNtew%2BNKL4T4uMzczJ6SEV3NKaiCWPQkzZEJTo1oAwDWVtn0wYxKce3TGVtRJ7GxNz4bYck4albiAcMiSTf1woHrd3PvVcG6t40RJEZmAQMq9QLnPyU0dzro6LvBBu49xiAAElsVA06A9HV4WztbOM48J7CGZcrsIqvilJk4ehkebe54gBiSQwfa/QTj8j3sb8OBTG2BwHEvBPDcF4KgTgZwzANQUDMOY6osQ8FIE6jgWg%2BGQxADiaIs/JC8DBFwDQ0RF9aFICvjgvAFAgGiAfo/pA4CwCQGgS8gqyAKAIA/8YgACQBiAuA9hpRogaBa0whX8IBghNBeBZJmBiAGwl9SBUDqgYQCxghtAUwD898/82BBAixYRkDSAsAqU2UbhX9MCsA9IjBxBD9eB8AxRnoSR6CH9MBVAUxngFg990ZShKDaA8BghiAcCXAsBKCN0LxOA98SRiANIlAGQmDgAxCjBkCpgqADBgAFBEx7R3wgFGAFDeB%2BBBARAxB2ApAZBBBFAVB1BWDSBdBGgDAtDTBzB9BxDX9IAphERyh6DvwqYAjvx6ASRrxXgPgfwqZa1q92oqAvQYjlg0AsBrIGBvxXRCAFAsE99UAlDCQsBfCoBmA2AQBjZEgBBSBeZO5OAMQNAMQeA%2BErBCCkgHAGBnBXB6gvAOjRguhChYh4hKjkgujBhBjMgkg%2Bj8hughhShWiBBWgagBgGgmh5iKgqgRg/BOhpiBjhhahRiVi9ipjxguAphN9Zh5gJAZ9OB59SB79l9OAV519kBlguAzIoCzINBs5cBCAylMQPhTjeAP9p8phDIzRug7IT8z99BOBL87jKCn8X83999tDriOAMQF8ETOAgTUTqj4C2jJAgA) a version that decomposes the State enum into a tagged union, with the tag in a `Cell` and the data in an `ActorCell` (in the editor rendered as a `RefCell`, but the conversion is straightforward).

**Comment by khoover** [comment]
[Update](https://rust.godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1DEArgoKkl9ZATwDKjdAGFUtEywYgAbAA5SjgBk8BkwAOXcAI0xiEABmLwAmUgAHVAVCOwYXNw9vP1T02wEgkPCWKJj4pKtMG0yhAiZiAmz3T19LTGsihgamghKwyOi4xMtG5tbcjvN%2BwbKK0YSASktUE2Jkdg4zTABqc3QQFC7aY4BSAHYAISdT0j2AJUwqO9paB%2BeqAFkTAiuACLnWLXc4aACCuwOBCOIDYLGO30MJjEtAAngDiKhksDQRCweDzglYucAKzXLDEPAAN0wECctAEmAeLmSaIeAMwERMwAeyn6eDE2AAjg8RcsyUD8cSyddiJhksQICYfBLSVLwYx3Hs%2BgR9lc8eC9sa9sp5clSASTU9MEx0OyrSaAFqocp4ZkEwGeiEmBiZPYApiNYFOQmxbB7A2O42KhUgPZIhgo94YrE42JOABqtRD11QqhD9oYewAYgwAPIMLYQCXh4HYeuWiHW%2BV2tHxxPJ9GY7EhsPYJuEy4agnmUw2PYASQYIWIffrkZu0b26CDTHjXzetBDgeDGf7jeXsz18a3Id1mEP%2BOH3vBeBYyW3%2B4X09n8/Di8N1qoxZCAHcaz2ABaBchC6KhP2Xa0wNoCCo2ba1EONY92D2Ldjn/CAL2OM0FVWKCkJXNcNxedCQEw3cmEghDCMI2Nkg7ZFUVTbEMMwADs2QNiAOWfCaNoxcAT4o1CK9fixMJfifz2PAFAAfQAL1dCJ3QgIkvCsKhlmAhcIlQVxqJEpCWCDZAEEwBQwEgTSADoUJs4BMAIGsHmwkAXTdTAJXEm98Sk4tZLk%2Bi1ISDTwO0kCPz0gz4KMxCTIIMyLKsiBbPsxznNWHVGhPEBcOSby4ok5dpKIOTW3tEKwtgh5qTEeNwW02LaPoAg9hYP4iMaSNYgBA5wJs1dGhsvTiCxP85I6zLcQI40ErM/rYLsnLMAcpzAOagTjTc/KeqBD9fQUJgqH1Jd%2BK25CBvspRnLcjyVK83E9gAeme0tfTqARmL2KiiQSVJZKZdA/uhINL3Oi7O2YnsGJARoAGs6XUqauqYGz6MKi6BJs4IysIaIa1mrG9hs/hiDk20zLUy4nCoK4nD2Kga1rL9ieNIa0YqtE9oTJiUxh7iIDq2gWaJ2i0pWuz1rc5421FiGkMBQcsZl217T2hdkkMPBkBSv7wRsEhjjKrnANEd5MHQPYCD/HXLwSFZlYuu7lPdDWPxcomJOtYr/Ot1BFNdpHQs0pqzri61WvazqOZ52yOZGkhxsmv5CZBInDuO07Wa2%2BaED2VKrsl80DGrF3PKy5HOoAKg5sOc%2BJrCVpw80OTXT89no%2BNhRMTBe8EpreoXKH%2BbTY50DTCAe77rynbZpuweOWX7Tb7qDT2Ln4yYQ3iAH92IxH7sx5ACfsQgbeiGIYS2eNBfcvu90Hjkwf9ojT2FcQ73RN8oqf5K4tRrJ1NupUOOkPzlmSD0EMXxfj/GfHWD8m1jKmXzhLMGa1MqGRVs3EAy9ubAlfjqV0dIYF/GOCZZIhcloJ0AagCaU0XKLicBzemexM4nSwTfKubVa5rhslzMWJpAS8TngJOS%2B89ihCZKIn2P9ZEjj9rQiawUQHhTARGCBUCMykLgVmHMGY8wFgzEWUsFYqx0lrA2BBViIxIPiigxaVBlroIyhtcO2DF55XNBIoQxCIA6PIUwSh8c%2BFKJTplB49MWE0zYQwI6HC7FY24XsXhw16KCONMI6%2BF1xEEIXFIkIMihFyOKSOP%2BEJ7yPnfBGGGjMSBThnNEapnDjTSVPpQ5JoDEkmijihOORcXFOXCWnBukcnKo36dQvhGVhnywjiadh2cMntQcalFaDw64tMbjtVuAZ27ry7nsae/dhESO6WzQ%2BLFYbtKnr3Xucyb6CWWdaO%2B7BcFq3ZHsteNwN4fK3jvPeeTEHuMeSaS5AsT6TwviQB5XCSnz3Lg9J%2BL8Fzv3mYreFn9MXFXKZCJQ0JYTRCxDEEA2AxokBmn5QkMpySUhpHSTk3JgBqg1MkEwEQ9hahYNlbETgmBQjsQ0bEyRLaiJLEwOglsICGMLGiYsZLiX1mycaAA0nQeg6BpX5llfK8lc5wzKoDGmEV6BBy%2BzvA%2BWgewFX1LJjy5IfKBVLlxVCQ4xwqAsAIBcG4AJZKPiYJ8m1u8mAKFLJ6hVDwSwkASnqOcN505UsqVa31Ch/XcztUKh1/L8V2Okh65yqiaqM09fGZJUbiAxqaRmMAHA5JKvUTaBQJhaC6MYSWcNY0FznJWYlVB4Etm0RglQY4maTUSPzTZP8VI9RyTHCFBIo7LZ/UNYO8CxwYZjqBRGCdU78azoIEqP6G6l0OxXYRIdxxxWSs1V5CRu69QpXzZEh2V6NV7FtgQfOUYbwOweLPZZF6QBqotjelFH572YEfZ659CRgNvo/V%2B51QJf2crPZk7Ff9cVso5WOEwE4DaXxlfA2xILjQMFQLYKg7Y0KnBDJozIuZtXGLlaYysZciBZt2JY%2BsV44rBFnPGV8VbQy8fNUm5pBGSBEZE8Chu2HGa/nYhAcjlHqP0YEIxoxzCWNljY3STNjqlDcYNQ2odA6TRme7daFTeAqOnlOILGzVG0OIX4yMBpAnyJKZc1/dDCi4rydKgHJSnkqqgMihoyBDGDFMe08WXT5im68uzY9ax5nLpLTc8QGyZUQsPRGUTWyTm0Q5aYIjQmPl/PWkC8WIJj40RBXNGF8CDxh0yUtaxxLEUFzqYYNU%2BuRNbOOJxo07LgVgoDY/taIbUdfFsCFmIF%2BjM3GjIurZLLOWA6m2FrCravmsUankbear7KFN7DCcAkOaiIt7F69Al4sDqldtI4hdbo3E7komqbTGpTjsmhq%2BdpOdDGsKma7Bbr4Cosae0Q9v4IZOKaZ1Z16sxmbHPdWxlpxG2wkTeXOak7HLSrRBYMEMGYOqAbPdEce1hnb3dpm%2BMubdJitLdsrloOK3lnFYgPgS2u2Dt48w/53DE4hCPkII8ZA0mhAPAAOro6PAegQoAaPvBDGYPAClLzhlEX%2BW08N7Nq4zBrrXvHrRjmV3JDmjEkzQzTOeM3Jo9dlat8RXmtvR69gzPLus15/MHolW1SXhiXC%2BgIKG3NxYLdGDkmgMP5OIe2NCmeY36RTcIIbtJZ38NY/rEEAnht6kU9OBN9rqxhp/7QixEYAvN3S/peG9H4Auf4/LAwRV3%2BVWTRZ/17Xhc9fu22Wzy3/PbfXE/b8391pAUqzkybwXwfRdq/N7j6PqWzkJbL5rHZIMGwgzBGb3aTVXBeKUs71Ps7WBkBz6VzXwtWkG9R1X21Ahjfb/AAKx/Ibz%2BeoM1L4icEAADQb3NyX0t2f23xuggB/yAj2BPzP1on23x27xn2vzkmzwXxe1AKWmHwgLbygKH173wN32IH3xjyPwgBP352QOnxXFqHJgwPv0m3RT2Cfzzxf16mGwwP52mwgh/xDDYTTzeW%2BCAJAJNEIJdzwPX2gPYOAjgJ4N%2B0q09FxXEwzBlz2B9wjGD3zFD0EFDQzXFwIG0K03UM0K2WkibxHwLSu3B0L2T1o1T01zLwb3v2WmX0Fy71oNwPYL70QXsKNxLyEIVw/lcOzw8OUP81dRhGOEgRJQKXCCN0NCiNhBMmIERhJWUAQEMCIBYEojPwJHkxFzajF1oAl2QHPDl2CICwPXjHiObSfCcBKLKOl0qOsVEWSCyMEFdHjEyOyNdDyIzGrhRiaKMKl1i1MJ419yHH81UMaNaIjBGMlwqI0KqIJzO0wib3jHUOz3jFl0Tx1H7W7UsywLohqL2EMWOHoDKy1XUC8wAkWLGILBOK2k2NVzODuMoJc1oh2LeMFhPyKRePf1d0aBty7CuUFiby%2BMImH2t3dzBIhUwmzwnwEiyRGyIDRSxg6L6IRFNE6JyLyJCIw08LOyy18KTy8AeJaJWNkwzjiSzg7lsliJshDXKheDcUO0UMkgC1Ox7xdw5jJMjFCllxcJsKx1G23xhLXHCKpWtEWTOxmUsKmjkg%2BnMmQERk1U/VklLVCmGOuy7R1M6nBXtzUNWMIjlPXmSQgGrk1IUHRgPWZPkliOZjcMt1jiQNxVlLpI4WkhmWHyVJVNqHVIgBtO1K8F1NsJu2SSNK9ycDMO7XNJ%2BUtOtIQFkjtOyxZKdN4knX12BN%2BiJIvwB3QDoSMFIKwAFJGNl312WLjOeOG1JLb2CDQO4IQPPUMMrLKw7liPjEZIPQeCxK6JxN6IHMoieSUL91vGSOOGxAUHXWiBeHyIqUtWrIXE5HlAggMNKNGOXJpP4gIDRBFT2AABUmgMoeYhAFy4o2k5ymYmC7CKSRTqoxTZxICgSWExzpjbxZiJiPwGQmQ6ld4HjtySNM9ixkBGQQhyzDjay3tnyGzZ990t8FDtooLrhO4zieyyBO48TujcTsSRz3SykZily1D5ijVsR/ydRDCliSLqTgLK8blOk1FF9MtxS28r8b9EKWykIhsYKCZ8D38Noh5OCNAxDjQEzULozrlJ5AMFSgT/SqxVSgzQ4WZRyWDsCny%2BLBp6D0CiCuKBIeKBp6zszriwM%2BoRKrMzSvTs44S7dWJIUz4ZKhk/S/hlSFLAypVlKnp9ssY3U4RMAER7LKFxLzj8x3UsQWByomAAIMKHSgoD1mYB49LEDMUjt3zzVCiD08Nii2yqyaKay5Mzi6iAjKTxj5jHcYxsLBzKqBinAhjOoSqTCyqpjzUiiND9chN9UnAa0mBSK7FXj1Juq3j1cgiddlwfiBqOAqJi9S9yr30czYSJqqJJKQxNCIjPziKurJrSLjDdDw8KL2z4YOqQxureqXsLCgSIDbzIz/CGiZqdyWDN9ldpSuTvxixvD48rr9SvBpqRq6KP4JD4ZnqxMNrvyFicqysgLzCZ9nzPqPwDqjrq1a1TqMdWDxlYjJknEmSMz4qkLYl4lrL4bRsGT39QyrTYj%2BLl8HhxrQoyaD029s8qb5q3d1Jaar5jKc83ThxVLUqXr/seTiwTBkhgBSzg5Hz9i7s1CqLyi8rJi/qWCo4ssMaRtYKkqTQIANt59B4IwNA29P1GBqYnBWELLjQNtGyOLldP81KLMpbOz0KBpYi%2BzKqejqr9kUrSlkSaC8b6SfSnKcz5Kko1SpUQzBSwz6q9S/DQ62plrvdTSkJgqkybS0zYrMz6ambvkOTJ9xzE0QbSLVyXgKKKzcq5jaKtk9yDzjziBTzX8fcK8/ZKQ2TYbyThTmKNKlRU6%2BSpS0qVCc6S7fyQgC7wbAaZb7rXq9gwKmRILYIG9eK26RsmzdKE0WDjjUKuzhsHasLsTna8LXaM7VLgaqlh6ak0wB7NyDrIbI8VxJ5GLbCW7lbNL2KdKTLVaTaIIZ6JSiD3YzLRKvaEkflJLx5pK10QBfS/aXKAzA7NVPKQRua1toi/KAqblgqLiQAqBwrIror7b7Tsa26w4hJn6YHM6PyqUMrxw2pyw/wQh0BJNOr%2BwsIpa%2Bwd5pN%2BxWsGAbikcEsUclUuHF6SGsq9hqHqk6HT6i7qGmHGwFM2HmN4szFOGDVuHrgOBVgzgOBSReBPBuBeBUBOBHgzA2oFB1hNh9RiQeBSAvUOAtAREDdYhYgbJrG7H7GHGvB9BOBJBeAEQuANANBSB1GtBSAtGOBeAZyvGzGLHSA4BYAkA0AHxJUyAKAZDon6AYhiAuBLgfAvGaAW1ogZyIAIhNBeAVJmBiB2wNHSACmmg0RywIhtBagzGTGom2BBBKx0Q8nSAsAmU%2BV3gZySmsATIjBxBzHeB8B5Q6haQunfHMBVBag/htgTHcYugWnSiIhSCimXAsAWmD17xOATHaRiA9IlBORengBSijA8nVgqADBgAFBMx3Q/wIFGAtneB%2BBBARAxB2ApAZBBBFAVB1ABnSBdAuB9BDBjBdH9A8AIgZzIBVhsQegumgIGZoWgJ6BaQnwARYhgIGYW1UACEqAgx0Wx6MBMBHIGAgJfRCAFACETHUAdmqQsAIWoBmA2AQBLYMgBBSBhZe5OAEgNAEgeBVhlNWB2AmgzI2WxAOW/wQ0WBYgEgSWGB4ZyMKGZW5XiyREahPpPAIBHAphPAAXAhgghhygRgAWCgWWshXA2g9BjWeh5hhgYgAXVWeg%2BhJgzXcg7Wugan6gJgBg9WFhDXxh%2BgtW9BZhmhrWDXbXVh9GNgtgJAlHOBVHvGWn/GN5dHkA4CbJUmbIRKIBcBCB6kiRYgT9eAQnFHVhzI7QRgXIrGbGHHq27GnHlHXH43fn/HAmQBgnTmwmYBEAQB1gCA2ULA4monkgYnQgBXOBTBzAU2uA02fAM3eBmXL48AjgAWnnhBzY3npAV2vm1AWn/nSAp0gkHmY2VG1GE3OByw/g%2B29hUAIJx3EpU303M2XAEnohBT83lhC3TnVhK3bGa3q3nGOAG2fHNHOAW222BmRFlGEgT2m2QPTHP22WsnMgQBJAgA%3D%3D%3D): now includes a port of `Rc` that allows for part of the data to be dropped only when the weak count goes to 0. So the `Deferred` and `LogId` can go in there while inner actor state and notify `Ret` can go in the strong data.

**Reply by khoover** [reply]
[Further update](https://rust.godbolt.org/z/aW8o6ds7W): moved the `SplitRcBox` creation/destruction to associated functions on the type, to keep it all in one place. Also, `weak_data` doesn't need to be `ManuallyDrop`, since it drops exactly when the `SplitRcBox` does, so got rid of that too.

**Comment by uazu** [comment]
I don't have time to go into all the detail with this right now. The pipebuf-limit code changes are taking all my (limited) time at the moment. 

Is this intended as another implementation option if the user doesn't require `forbid(unsafe_code)`? I did create `unsized_enum` crate to help with this part of Stakker, but I didn't want to impose so much `unsafe` for review on Stakker users.

**Reply by khoover** [reply]
Roughly, yeah; this is meant to solve that problem where there's the inner enum inside the ActorCell along with a portion of the reference count that holds the enum state.

**Comment by ducharmemp** [comment]
Would this (in theory) unlock some of the functionality that's currently blocked by the stabilization of `CoerceUnsized` and friends? I ask because `unsized_enum` has been mentioned in similar threads. I'm curious as to if this could get closer to having `ActorOwn<dyn Foo> as ActorOwn<dyn Bar>` type functionality, similar to what is available with `std::rc::Rc`

**Reply by khoover** [reply]
Maybe? I mostly wrote this to eliminate the duplication of enum state into the ref count. Then the idea of SplitRc is there's the strong data that drops when the strong count goes to 0 (i.e. the actor state) and weak data that drops when the weak count goes to 0 (i.e. anything that can be used from a zombie state).
